common/djangoapps/course_modes/views.py

```diff
@@ -241,18 +241,36 @@ def post(self, request, course_id):
         allowed_modes = CourseMode.modes_for_course_dict(course_key)
         if requested_mode not in allowed_modes:
             return HttpResponseBadRequest(_("Enrollment mode not supported"))
-
-        if requested_mode == 'audit':
+        if requested_mode in {'audit', 'honor'}:
             # If the learner has arrived at this screen via the traditional enrollment workflow,
             # then they should already be enrolled in an audit mode for the course, assuming one has
             # been configured.  However, alternative enrollment workflows have been introduced into the
             # system, such as third-party discovery.  These workflows result in learners arriving
             # directly at this screen, and they will not necessarily be pre-enrolled in the audit mode.
-            CourseEnrollment.enroll(request.user, course_key, CourseMode.AUDIT)
-            return redirect(reverse('dashboard'))
-
-        if requested_mode == 'honor':
             CourseEnrollment.enroll(user, course_key, mode=requested_mode)
+            enterprise_learner_data = enterprise_api.get_enterprise_learner_data(site=request.site, user=user)
```

Comment:
@haikuginger It would be worth adding some comments to this block that explain the added logic (or rather, mention why it's necessary).